include/audio_backend/__asio_audio_device.h

```diff
@@ -0,0 +1,527 @@
+    {
+    case kAsioSupportsTimeInfo: return ASIOTrue;
+    case kAsioEngineVersion: return 2;
+    case kAsioSelectorSupported: return ASIOFalse;
+    default: return ASIOFalse;
+    }
+  }
+
+  static ASIOTime* buffer_switch_time_info (ASIOTime* time, long index, ASIOBool /*direct_process*/) {
+    return _instance->on_buffer_switch(time, index);
+  }
+
+  ASIOTime* on_buffer_switch(ASIOTime* time, long index) {
+
+    _fill_input_buffers(index);
+    _user_callback(*this, _io);
+    _fill_output_buffers(index);
+
+    time->timeInfo.flags = kSystemTimeValid | kSamplePositionValid;
+    time->timeInfo.samplePosition = _sample_position;
+    _sample_position += _buffer_size;
+    time->timeInfo.systemTime = timeGetTime() * 1'000'000;
+    return time;
+  }
+
+  CComPtr<IASIO> _asio;
+  const string _name;
+  device_id_t _id;
+  long _num_inputs = 0;
+  long _num_outputs = 0;
+  vector<sample_rate_t> _sample_rates;
+  vector<buffer_size_t> _buffer_sizes;
+  buffer_size_t _buffer_size = 0;
+  bool _running = false;
+
+  ASIOSampleType _sample_type;
+  ASIOSamples _sample_position = 0;
+
+  using __asio_callback_t = function<void(audio_device&, audio_device_io<float>&)>;
+  __asio_callback_t _user_callback;
+  audio_device_io<__asio_common_sample_type> _io;
+  vector<float> _input_samples;
+  vector<float> _output_samples;
+
+  using fill_buffers = function<void(long index)>;
+  fill_buffers _fill_input_buffers;
+  fill_buffers _fill_output_buffers;
+
+  vector<ASIOBufferInfo> _asio_buffers;
+  static audio_device* _instance;
+  ASIOCallbacks _asio_callbacks;
+};
+
+audio_device* audio_device::_instance = nullptr;
+
+class audio_device_list : public forward_list<audio_device> {
+};
+
+class __reg_key_reader final
+{
+public:
+  __reg_key_reader(HKEY key, const string& subkey) {
+    const auto result = RegOpenKeyExA(key, subkey.c_str(), 0, KEY_READ, &_key);
+
+    if (ERROR_SUCCESS != result) {
+      throw runtime_error("Failed to read info from registry: 0x" + result);
+    }
+  }
+
+  ~__reg_key_reader() {
+    if (_key) {
+      RegCloseKey(_key);
+    }
+  }
+
+  vector<string> subkeys() const {
+    constexpr int max_asio_name_length{32};
+    char name[max_asio_name_length];
+    DWORD size{max_asio_name_length};
+    vector<string> keys;
+
+    for (DWORD index = 0; ; ++index) {
+      const auto result = RegEnumKeyEx(_key, index, name, &size, nullptr, nullptr, nullptr, nullptr);
+
+      if (result != ERROR_SUCCESS) {
+        break;
+      }
+
+      keys.emplace_back(name);
+    }
+
+    return keys;
+  }
+
+  __reg_key_reader subkey(const string& name)
+  {
+    return {_key, name};
+  }
+
+  string value(const char* name) const {
+    DWORD size{0};
+    RegQueryValueExA(_key, name, nullptr, nullptr, nullptr, &size);
+
+    string value(size + 1, 0);
+    RegQueryValueExA(_key, name, nullptr, nullptr, LPBYTE(value.data()), &size);
+
+    return value;
+  }
+
+private:
+  HKEY _key{0};
+};
+
+class __asio_devices {
```

Comment:
Most of the functions in this class can be `const`.

Reply:
Fixed in ce687e8